### PR TITLE
optimize process-user-data startup time

### PR DESCRIPTION
**Pull request by mkulke**
the binary is currently importing a lot of libraries which have costly init() procedures (around 8-10s at startup) this is due to sharing some consts with the CAA modules. We can avoid that by inlining code in process-user-data and moving shared consts into dedicated modules.

We can use a cpuid field to narrow down a hypervisor, so we don't have to probe an IMDS endpoint while circling through the cloud provider, delaying startup.

this is the result w/ debug enabled, the crypto libraries that are used to hash initdata are still consuming a few, but it's far better now.

```bash
Nov 19 15:51:19 fedora process-user-data[676]: init main @1029 ms, 0.039 ms clock, 2600 bytes, 22 allocs
```

### CODE REVIEW

File: src/cloud-api-adaptor/pkg/userdata/heuristics.go
Comment (stevenhorsman) [inline src/cloud-api-adaptor/pkg/userdata/heuristics.go:1]: Is it worth mentioning this file in `addnewprovider.md` as it's a bit less obvious that the cloud-providers changes?

File: src/cloud-api-adaptor/pkg/paths/paths.go [hunk 1]
```diff
@@ -0,0 +1,11 @@
+package paths
+
+const (
+	AACfgPath          = "/run/peerpod/aa.toml"
+	AuthFilePath       = "/run/peerpod/auth.json"
+	CDHCfgPath         = "/run/peerpod/cdh.toml"
+	InitDataPath       = "/run/peerpod/initdata"
+	AgentCfgPath       = "/run/peerpod/agent-config.toml"
+	ForwarderCfgPath   = "/run/peerpod/daemon.json"
+	DockerUserDataPath = "/peerpod/userdata.json"
```
Comment (stevenhorsman) [inline src/cloud-api-adaptor/pkg/paths/paths.go:10]: Is there an opportunity to move the existing constants to use this instead e.g. in cloud.go. I'm slightly concerned about us now having two copies that could go out of sync?

File: src/cloud-providers/docker/provider.go [hunk 2]
```diff
@@ -72,11 +72,11 @@ func (p *dockerProvider) CreateInstance(ctx context.Context, podName, sandboxID
 
 	// Create volume binding for the container
 
-	// mount userdata to DockerUserDataUrl=/peerpods/userdata.json
+	// mount userdata to /peerpods/userdata.json
 	// This file will be read by process-user-data and daemon.json will be written to
 	// /run/peerpods/daemon.json at runtime
 	volumeBinding := []string{
-		fmt.Sprintf("%s:%s", instanceUserdataFile, DockerUserDataUrl),
+		fmt.Sprintf("%s:%s", instanceUserdataFile, "/peerpods/userdata.json"),
```
Comment (stevenhorsman) [inline src/cloud-providers/docker/provider.go:79]: I guess we don't want to pull in the `DockerUserDataPath` from paths and we want have the cloud-providers->CAA cyclic dependency. Maybe a comment here and in paths.go to say they have to stay in sync is the least bad approach then?

File: src/cloud-api-adaptor/pkg/paths/paths.go (hunk 1 above)
Comment (stevenhorsman) [reply src/cloud-api-adaptor/pkg/paths/paths.go:10]: I mean these: https://github.com/confidential-containers/cloud-api-adaptor/blob/a22b492ebc709cac8b361e850647e67494d7b973/src/cloud-api-adaptor/pkg/adaptor/cloud/cloud.go#L38-L41. They are in the cloud-api-adaptor module, so I don't think there is a dependency problem?

Comment (stevenhorsman) [reply src/cloud-api-adaptor/pkg/paths/paths.go:10]: If we aren't using them apart from here can we just delete them?

Comment (mkulke) [reply src/cloud-api-adaptor/pkg/paths/paths.go:10]: that's true. we are probably using them when creating the user-data file, I probably mixed something up, in theory those should be able to use the paths module

File: src/cloud-api-adaptor/pkg/userdata/heuristics.go
Comment (mkulke) [reply src/cloud-api-adaptor/pkg/userdata/heuristics.go:1]: actually it's more, you also need to implement a way to retrieve userdata for a new provider

**Review by stevenhorsman — APPROVED** [review]
LGTM. Thanks for the updates!

File: src/cloud-api-adaptor/pkg/userdata/provision.go [hunk 3]
```diff
@@ -78,51 +83,57 @@ func (d DefaultRetry) GetRetryDelay() time.Duration {
...
 }
 
 type AWSUserDataProvider struct{ DefaultRetry }
 
 func (a AWSUserDataProvider) GetUserData(ctx context.Context) ([]byte, error) {
-	url := aws.AWSUserDataImdsUrl
+	url := AWSUserDataImdsUrl
 	logger.Printf("provider: AWS, userDataUrl: %s\n", url)
-	return aws.GetUserData(ctx, url)
+	return imdsGet(ctx, url, false, nil)
```
Comment (bpradipt) [inline src/cloud-api-adaptor/pkg/userdata/provision.go:96]: shouldn't the bool be true here ?

Comment (mkulke) [reply src/cloud-api-adaptor/pkg/userdata/provision.go:96]: aws doesn't encode the userdata with base64, does it? at least it wasn't in the original code

Comment (bpradipt) [reply src/cloud-api-adaptor/pkg/userdata/provision.go:96]: Correct.. I forgot about it.. Do you mind adding a comment above the code ?

**Review by bpradipt — APPROVED** [review]
/lgtm

**Pull request merged**